### PR TITLE
Reduce some diffs in exec_copyout_strings

### PR DESCRIPTION
**Pull request by bsdjhb**
Originally I was going to try to look for places where we could remove casts due to the builtins working with uintcap_t, but exec_copyout_strings() was the only place I found.

### CODE REVIEW

File: sys/compat/freebsd64/freebsd64_misc.c [hunk 1]
```diff
@@ -691,19 +697,19 @@ freebsd64_copyout_strings(struct image_params *imgp, uintcap_t *stack_base)
 	 * Allocate room for the argument and environment strings.
 	 */
 	destp -= ARG_MAX - imgp->args->stringspace;
-	destp = __builtin_align_down(destp, sizeof(uint64_t));
-	ustringp = destp;
+	destp = rounddown2(destp, sizeof(uint64_t));
+	ustringp = cheri_setbounds(destp, ARG_MAX - imgp->args->stringspace);
```
Comment (qwattash) [inline sys/compat/freebsd64/freebsd64_misc.c:701]: Is this missing in the purecap kernel as well?

File: sys/compat/freebsd64/freebsd64_misc.c [hunk 2]
```diff
@@ -632,11 +633,12 @@ freebsd64_copyout_strings(struct image_params *imgp, uintcap_t *stack_base)
...
 	    ssiz + stack_offset));
-	destp = cheri_capability_build_user_data(CHERI_CAP_USER_DATA_PERMS,
-	    rounded_stack_vaddr,
+	destp = (uintcap_t)cheri_capability_build_user_data(
+	    CHERI_CAP_USER_DATA_PERMS, rounded_stack_vaddr,
 	    CHERI_REPRESENTABLE_LENGTH(ssiz + stack_offset), stack_offset);
 	destp = cheri_setaddress(destp, p->p_sysent->sv_psstrings);
-	arginfo = (void * __capability)cheri_setbounds(destp, sizeof(*arginfo));
+	arginfo = (struct freebsd64_ps_strings * __capability)cheri_setbounds(
+	    destp, sizeof(*arginfo));
```
Comment (brooksdavis) [inline sys/compat/freebsd64/freebsd64_misc.c:641]:
```suggestion
	arginfo = (struct freebsd64_ps_strings * __capability)
	    cheri_setbounds_exact(destp, sizeof(*arginfo));
```

File: sys/compat/freebsd64/freebsd64_misc.c [hunk 3]
```diff
@@ -659,19 +662,22 @@ freebsd64_copyout_strings(struct image_params *imgp, uintcap_t *stack_base)
 	 */
 	if (execpath_len != 0) {
 		destp -= execpath_len;
-		imgp->execpathp = cheri_setbounds(destp, execpath_len);
-		error = copyout(imgp->execpath, destp, execpath_len);
+		destp = __builtin_align_down(destp, sizeof(uint64_t));
+		imgp->execpathp = (void * __capability)
+		    cheri_setbounds(destp, execpath_len);
```
Comment (brooksdavis) [inline sys/compat/freebsd64/freebsd64_misc.c:667]: Maybe cher_setbounds_exact here as well? I think it should be safe with PATH_MAX=1024.

**Review by brooksdavis — COMMENTED** [review]
I think most of the cheri_setbounds should be cheri_setbounds_exact. In most cases (e.g. canary, pagesz array, etc), something has gone quite wrong if that results in a failure.

File: sys/kern/kern_exec.c [hunk 4]
```diff
@@ -1840,7 +1834,12 @@ exec_copyout_strings(struct image_params *imgp, uintcap_t *stack_base)
 	if (imgp->auxargs) {
 		vectp++;
 		error = imgp->sysent->sv_copyout_auxargs(imgp,
-		    (uintcap_t)vectp);
+#if __has_feature(capabilities)
+		    (uintcap_t)cheri_setbounds(vectp,
```
[the remark this answers is not in the document]
Comment (bsdjhb) [reply sys/kern/kern_exec.c:1838]: Oh, I missed it when rototilling all this code during the hybrid RISC-V kernel bring up preparations as this isn't purecap-kernel specific at all.  I added the sv_copyout_auxargs() hook around that time.

File: sys/compat/freebsd64/freebsd64_misc.c (hunk 1 above)
Comment (bsdjhb) [reply sys/compat/freebsd64/freebsd64_misc.c:701]: I haven't looked, but I'll merge dev to the purecap branch again "soon"

**Comment by bsdjhb** [comment]
I will add a followup commit to use setboundsexact as many places as possible.

**Review by brooksdavis — APPROVED** [review]

**Pull request merged**